Report v2: check/x boolean audits, expand details failing audits, optimal val

ebidel opened the pull request:
R: all

Before: 

![screen shot 2017-04-04 at 12 51 49 pm](https://cloud.githubusercontent.com/assets/238208/24675857/83d50056-1935-11e7-939e-d0fcdd1eedd7.png)

After: 

![screen shot 2017-04-04 at 12 49 07 pm](https://cloud.githubusercontent.com/assets/238208/24675726/1e6583da-1935-11e7-9b26-3fb4af705191.png)

- Expands failing audits + expander arrows
- Support for boolean audits with X/check
- Renders `optimalValue` text on audit title. Final UI we'll need to work out.

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -98,19 +102,42 @@ window.ReportRenderer = class ReportRenderer {
+    const isAudit = 'result' in auditOrCategory;
+    const tagName = isAudit ? 'details' : 'div';
+
+    // Score template. Audits get a details dropdown. Categories don't.
```
if we wanted to, we could make the whole category collapsable. For now, I decide to keep it as v1.

patrickhulce commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -215,12 +243,15 @@ window.ReportRenderer = class ReportRenderer {
     }
+
+    element.appendChild(
+        this._renderScore(audit.result.score, title, audit.result.helpText, audit));
```
why can't we stick to `audit.score`?

ebidel replied:
That doesn't have boolean information. Booleans come back `score: 0`

patrickhulce replied:
Yeah moved discussion to https://github.com/GoogleChrome/lighthouse/pull/1963#discussion_r109763756 :)

patrickhulce commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -31,6 +31,10 @@ const RATINGS = {
  * @return {string}
  */
 function calculateRating(value) {
+  if (typeof value === 'boolean') {
```
:( can we just keep that logic in the report generator?

patrickhulce commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -98,19 +102,42 @@ window.ReportRenderer = class ReportRenderer {
+
+    // Score template. Audits get a details dropdown. Categories don't.
+    const tmpContainer = this._createElement('div');
+    tmpContainer.innerHTML = `<div class="lighthouse-score">
```
😱

ebidel replied:
Don't be scared. Completely safe since we control all inputs for the initial set.

patrickhulce commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -98,19 +102,42 @@ window.ReportRenderer = class ReportRenderer {
    */
-  _renderScore(score, title, description) {
-    const element = this._createElement('div', 'lighthouse-score');
+  _renderScore(score, title, description, auditOrCategory) {
```
the doctype lies! since we're shifting away from displaying high precision scores is this still necessary? what about a `precision: binary|grade|numeric` property or something for each audit and we can say goodbye to `typeof`

ebidel replied:
Score could also be an object with value/precision props. Or maybe`type` for the latter. Examples:

```
category.score = {value: 100, precision: 'numeric'}
audit.score = {value: true, precision: 'binary'}
audit.score = {value: 'A', precision: 'grade'}
```

brendankenny replied:
can someone sum up what was decided? My thoughts:

- audits are going to score in some fashion, whether that's 0-100, bucketed into grades, or pass/fail. Even if we require all audits to report a score as a number, it's still going to be an integer in that interval, buckets of every 10ish numbers, or 0/100. It makes some sense then for the audit to declare how it scored (perhaps on audit.meta), even if the reporter changes how the score is presented.

- If on the other hand (or in addition to the above), we allow specifying the precision type in the config (on one of the audits in `categories.categoryName.audits`) it seems like we'd need some way of converting the score reported by the audit into the new format. `binary` -> `numeric` is easy, but (as discussed), what's the true/false cutoff for `numeric` -> `binary` and where do you specify it?

patrickhulce replied:
Here's my take on what was discussed

Immediate actions: 
* add a `scorePrecision` or similarly named property to the performance audits `meta` object specifying a `numeric` precision
* add default `scorePrecision` of `binary` to the result
* update ReportRenderer to use the `audit.result.scorePrecision` combined with `audit.score` to decide how to display (no more `typeof score === 'boolean'` outside of ReportGenerator)

Potential future course of action that need further discussion: We remove `score` from the audit result object and instead have them return values that are compared to the declared target/point of diminishing returns/median values and a score is generated for them, boolean audits return whether they pass or fail as they do today and a score of 100/0 is generated for them. We explore the possibility of introducing letter grades and changing the score curves.

brendankenny replied:
> add a `scorePrecision` or similarly named property to the performance audits meta object specifying a numeric precision

should be noted that is basically just giving `expectedValue` a new name :)

It will be nice to always have a numeric score, though.

patrickhulce replied:
if you mean renaming a property to align with it's meaning and changing the types of values it can hold then yes it's basically that :)

brendankenny replied:
an enum is an enum :P

ebidel replied:
@patrickhulce do you want to craft those changes in a PR? I am working on some other stuff with `<template>` and can rebase when that lands.

patrickhulce replied:
@ebidel sure thing

patrickhulce replied:
#1967

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -93,24 +116,73 @@ window.ReportRenderer = class ReportRenderer {
+   */
+  _renderScore(score, title, description, auditOrCategory) {
+    const isAudit = 'result' in auditOrCategory;
+    const precision = isAudit ? 'binary' : 'category'; // TODO(ericbidelman): placeholder
```
fyi @patrickhulce

ebidel commented:
Removed `.innerHTML`! 

It's been replaced by templates.html which includes declared html partials (using `<template>`). It will also be the future home for ported handlebars formatters and other UI scaffolding. Templates get added to `report-template.html` using our injection technique. It's unclear how devtools will bring in templates.html.

This PR also introduces a `DOM` class to organize DOM-related code. It includes wrapper methods for setting `.textContent` and `.classList` so we never run into errors with querySelector'd returning `null`. Plan to add some renderer tests.

The `typeof === 'boolean'` stuff is also removed. I've preemptively hardcoded `precision` placeholders for audit and cateogry scores. For now, `_renderAuditScore` and `_renderCategoryScore` construct an object: `score = {value: audit.score, precision: 'numeric'}`. The expectation is that `report-generator.js` will setup this object when it creates results and we can remove those lines.

PTAL.

ebidel commented:
Now with `scoringMode`. PTAL

paulirish commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -93,24 +116,81 @@ window.ReportRenderer = class ReportRenderer {
-    column.appendChild(
-      this._createElement('div', 'lighthouse-score__description')).textContent = description;
+    // Fill in the blanks.
+    const value = tmpl.querySelector('.lighthouse-score__value');
```
s/value/elem/

ebidel replied:
when for `valueEl` so disambiguate.

ebidel commented:
Added tests. I'm done.

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -64,15 +71,31 @@ window.ReportRenderer = class ReportRenderer {
-    } catch (e) {
-      return this._renderException(e);
+  static setText(element, text) {
+    if (element) {
```
if we're worried about these, should they just throw? Seems like it's definitely an error since we're creating the templates *and* the selectors looking within them

ebidel replied:
@paulirish asked for them so that the fails gracefully if a node cant be found. My pref would be to remove them altogether. Tests should really be catching missing nodes in templates.html (markup we control). It would also get rid of the `DOM` and simplify render methods a bit if we just set things directly (`.classList`, `.textContent`).

I'm also ok keeping.

brendankenny replied:
> @paulirish asked for them so that the report fails gracefully if a node cant be found.

I think @paulirish was asking before we had completely settled on the approach used here, so it may be less of an issue. Queries failing fast (as soon as we try to set `textContent` or add to the `classList`) is a good thing, I think. I also like the idea of moving some of the query verification to tests, but if setting `element.textContent` throws a TypeError, just testing e.g. `_renderAuditScore` will verify the relevant elements exist if it doesn't throw :)

brendankenny replied:
still think we should remove. Earlier (and non-silent) failures are better failures.

ebidel replied:
@paulirish says, "I am low stress on this. Do whateva :)"

Note to future me: give PI 24hrs and he will be low stress.

ebidel replied:
done

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -199,8 +279,10 @@ window.ReportRenderer = class ReportRenderer {
-    element.appendChild(this._renderScore(category.score, category.name, category.description));
+    const element = this._dom.createElement('div', 'lighthouse-category');
+    const score = {value: Math.round(category.score), scoringMode: 'numeric'};
+    element.appendChild(
```
`_renderCategoryScore` only takes three arguments

ebidel replied:
done

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -212,19 +294,35 @@ window.ReportRenderer = class ReportRenderer {
-      element.appendChild(this._renderDetails(audit.result.details));
+
+    const score = {value: audit.score, scoringMode: audit.result.scoringMode};
+    element.appendChild(this._renderAuditScore(score, title, audit.result.helpText, audit));
```
also doesn't need the fourth argument

ebidel replied:
done

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -81,36 +104,93 @@ window.ReportRenderer = class ReportRenderer {
+   * @param {!string} description
+   * @return {!Element}
+   */
+  _populateScore(tmpl, score, title, description) {
```
should this (and the render functions below) be called something like `_populateResult` instead of `_populateScore` since they include the score, title, and description of the audit/category?

ebidel replied:
done

ebidel replied:
done

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -199,8 +279,10 @@ window.ReportRenderer = class ReportRenderer {
-    const element = this._createElement('div', 'lighthouse-category');
-    element.appendChild(this._renderScore(category.score, category.name, category.description));
+    const element = this._dom.createElement('div', 'lighthouse-category');
+    const score = {value: Math.round(category.score), scoringMode: 'numeric'};
```
any reason to have this as an object? Seems fine as separate arguments to `renderCategoryScore`/`renderAuditScore` (and you may not even need `scoringMode` in there, at least not yet)

brendankenny replied:
just kidding, you do need `scoringMode` for the class name, but still seems like they should be separate parameters :)

ebidel replied:
done

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+'use strict';
+
+/* eslint-env mocha */
+/* global document self */
```
`/* eslint-env mocha, browser */`?

ebidel replied:
done

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+const sampleResults = require('../../results/sample_v2.json');
+
+const TEMPLATES_FILE = fs.readFileSync(
+    path.join(__dirname, '../../../report/v2/templates.html'), 'utf8');
```
don't need to joinnnn

ebidel replied:
done

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+    path.join(__dirname, '../../../report/v2/templates.html'), 'utf8');
+
+function setupJsDomGlobals() {
+  global.document = jsdom.jsdom(TEMPLATES_FILE);
```
shouldn't this test in the full context of the `report-template.html` page (with assets inlined)?

ebidel replied:
Right. I was just testing template output and the render methods in this file, not how they fit into the entire app. Using the whole page would need `ReportGeneratorV2`, which seems more end to end.

What I've done: added are more tests to report-generator-test.js so we know all the templates get injected.

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+require('../../../report/v2/report-renderer.js');
+
+describe('DOM', () => {
+  const window = self;
```
`window` appears unused?

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+    });
+
+    it('does not fail when given null element argument', () => {
+      assert.doesNotThrow(() => {
```
can drop the surrounding `assert.doesNotThrow` and just execute it. If it throws the test will fail

ebidel replied:
I kind of like being explicit.

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+    });
+
+    it('does not fail when given null element argument', () => {
+      assert.doesNotThrow(() => {
```
don't need `doesNotThrow` wrapper

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+    it('fails when template cannot be found', () => {
+      const dom = new DOM(document);
+      assert.throws(() => {
+        const clone = dom.cloneTemplate('#unknown-selector'); // eslint-disable-line no-unused-vars
```
can you just not assign to a var to lose the eslint directive?

ebidel replied:
sloppy...but done!

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+});
+
+describe('ReportRenderer V2', () => {
+  const window = self;
```
isn't the global `window` already set to `self`? Or...something? The globals are confusing here (and only used to access `ReportRenderer`)

brendankenny commented on lighthouse-core/test/report/v2/report-renderer-test.js:
```diff
@@ -0,0 +1,177 @@
+
+    it('should render an exception for invalid input', () => {
+      const renderer = new ReportRenderer(document);
+      const output = renderer.renderReport({});
```
Since we're not really validating anything right now this relies on implementation details. What about something explicit like
```js
const output = renderer.renderReport({
  get reportCategories() { throw new Error() }
});
```

ebidel replied:
done

brendankenny commented:
> there's stuff i could bikeshed.. like the DOM library API and how the templates are added to the page. But on the whole I'm happy here. Let's keep it moving.

If it really is just bikeshedding stuff that's fine, but this is the kind of thing that gets written once and we're stuck with from then on unless it bothers you so much in the future you file a bug/change it.

Now's the time to bring it up, even if we get to it in a followup :)

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -15,47 +15,54 @@
- * Dummy text for ensuring report robustness: </script> pre$`post %%LIGHTHOUSE_JSON%%
- */
-window.ReportRenderer = class ReportRenderer {
+class DOM {
```
feels like something I know ahead of time you're going to disagree with me on :) but I don't personally see the need for a separate `DOM` class (yet, at least), especially because the split with `ReportRenderer` doesn't seem completely clear. Anything wrong with these living there?

ebidel replied:
I knew you were going to bring this up! I wasn't convinced adding a bunch of generic methods to ReportRenderer made sense either. They could live as standalone helper functions, but devtools may need to extend them later.

ebidel replied:
removed

ebidel commented on lighthouse-core/report/v2/report-generator.js:
```diff
@@ -97,11 +98,14 @@ class ReportGeneratorV2 {
     const sanitizedJson = JSON.stringify(reportAsJson).replace(/</g, '\\u003c');
     const sanitizedJavascript = REPORT_JAVASCRIPT.replace(/<\//g, '\\u003c/');
+    // Remove script in templates files just to be safe.
+    const sanitizedTemplates = REPORT_TEMPLATES.replace(/<script>([\s\S]*?)<\/script>/g, '');
```
I agree with you. I recall that `el.innerHTML = '<div>just a string</div>';` was deemed unsafe and that's our own markup :) The argument from _someonnnnne_ was that we may add things later and not realize the mistakes.

brendankenny replied:
I think this is different enough, though, because it's not markup assembled in (potentially) multiple places, it's all in one spot. So I think it's better on that front. I also don't like calling it sanitized unless it is actually sanitized :)

brendankenny replied:
really think we should remove this, or call it "templatesWithScriptTagsRemoves" or something. Current version is easily worked around with an attribute on the script tag...I don't think this is an arms race we want to even start.

ebidel replied:
done

patrickhulce commented on lighthouse-core/report/v2/report-generator.js:
```diff
@@ -23,6 +23,7 @@ const REPORT_TEMPLATE = fs.readFileSync(path.join(__dirname, './report-template.
 // TODO: Setup a gulp pipeline to concat and minify the renderer files?
 const REPORT_JAVASCRIPT = fs.readFileSync(path.join(__dirname, './report-renderer.js'), 'utf8');
 const REPORT_CSS = fs.readFileSync(path.join(__dirname, './report-styles.css'), 'utf8');
+const REPORT_TEMPLATES = fs.readFileSync(path.join(__dirname, './templates.html'), 'utf8');
```
does that work just fine on windows?

ebidel replied:
done

brendankenny replied:
> does that work just fine on windows?

Yeah, Windows understands / and the `fs` methods will take care of the rest. The only real worry (that I know of) is if you want to specify an absolute path from the root, which is different cross platform, but we're using `__dirname` so it's not an issue here

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -81,36 +104,93 @@ window.ReportRenderer = class ReportRenderer {
-   * @param {string} title
-   * @param {string} description
+   * @param {!string} selector
+   * @return {!DocumentFragment} A clone of the template content.
```
Technically a doc frag, which is not a full element. We use it like an `!Element` though.

brendankenny replied:
But it isn't just grabbing the `content`, it also calls `document.importNode` which appears to return a Node.

ebidel replied:
TBH, it's a terrible API and name. `importNode()` returns the node you import. Since we're importing a `DocumentFragment`, we get back a `Node` of that type. 

![screen shot 2017-04-06 at 12 35 37 pm](https://cloud.githubusercontent.com/assets/238208/24772285/8e68f978-1ac5-11e7-85d9-923aa5da73a1.png)

brendankenny replied:
> `importNode()` returns the node you import

ah, I see. Should `_populateResult` take a `!DocumentFragment`, then?

patrickhulce commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -15,47 +15,54 @@
  * Convert a score to a rating label.
- * @param {number} value
- * @return {string}
+ * @param {!number} score
```
idk though @ebidel the great `Object<string, string>` debacle haunts us throughout our codebase now after listenin' to this guy 😉

ebidel replied:
good to know. removed.

brendankenny replied:
well, TECHNICALLY, `Object<string>` still works, Closure just won't type check it :)

ebidel commented on docs/hacking-tips.md:
````diff
@@ -13,9 +13,18 @@ Use [`--trace-warnings`](https://medium.com/@jasnell/introducing-process-warning
+
+```sh
+npm run start -- --output=json --output-path=lighthouse-core/test/results/sample_v2.json http://localhost:8080/dobetterweb/dbw_tester.html
+```
````
done

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -15,47 +15,54 @@
+    return rating;
+  }
+
+  // Treat as binary by default.
```
Right, good call. Left over from before. Removed.

brendankenny replied:
commit pushed yet?

brendankenny replied:
jk you haven't pushed yet :)

ebidel replied:
Just pushed the first set of changes. Updating tests to come.

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -64,15 +71,31 @@ window.ReportRenderer = class ReportRenderer {
   /**
-   * @param {!ReportJSON} report
-   * @return {!Element}
+   * Sets the text content of an element, safely. If the element does not exist,
```
done

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -64,15 +71,31 @@ window.ReportRenderer = class ReportRenderer {
+   * Adds a class to an element, safely. If the element does not exist, results
+   * in a noop.
+   * @param {Element} element
+   * @param {...!string} classes
```
done

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -81,36 +104,93 @@ window.ReportRenderer = class ReportRenderer {
-    for (const [key, val] of Object.entries(attrs)) {
-      element.setAttribute(key, val);
-    }
+    Object.keys(attrs).forEach(key => {
```
+1

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -81,36 +104,93 @@ window.ReportRenderer = class ReportRenderer {
+    const valueEl = tmpl.querySelector('.lighthouse-score__value');
+    DOM.setText(valueEl, formatNumber(score.value));
+    DOM.addClass(valueEl, `lighthouse-score__value--${rating}`,
+                          `lighthouse-score__value--${score.scoringMode}`);
```
done

ebidel commented:
PTAL. I think I got everything.

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -81,39 +96,104 @@ window.ReportRenderer = class ReportRenderer {
+  _renderAuditScore(audit) {
+    const tmpl = this._dom.cloneTemplate('#tmpl-lighthouse-audit-score');
+
+    const scoringMode = audit.result.scoringMode;
```
How we've gotten here has made sense, but it is weird that `scoringMode` is on `audit.result` but the score is on `audit` directly

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -81,39 +96,104 @@ window.ReportRenderer = class ReportRenderer {
+  }
+
+  /**
+   * @param {!DocumentFragment|Element} element DOM node to populate with values.
```
this is only ever called with a `DocumentFragment`, though

ebidel replied:
yep, but it support both

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -15,26 +15,31 @@
+  PASS: {label: 'pass', minScore: 75},
   AVERAGE: {label: 'average', minScore: 45},
-  POOR: {label: 'poor'}
+  FAIL: {label: 'fail'}
```
does PASS/FAIL work here? Seems weird to have a three state pass/fail system :)

ebidel replied:
Not really a pass/fail system. Just for labeling/coloring right now. We'll probably need to expand in the future if we support things like grades. Seems fine for now.

brendankenny replied:
> Not really a pass/fail system. Just for labeling/coloring right now.

yeah, I was mostly just commenting on the names since PASS/FAIL is usually a binary state. Fine for now

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -81,39 +96,104 @@ window.ReportRenderer = class ReportRenderer {
+   * @param {!CategoryJSON} category
+   * @return {!Element}
+   */
+  _renderCategoryScore(category) {
```
I'm actually going to change "_populateResult" back to `_populateScore`. It's populating the score dom with values.

brendankenny replied:
> I'm actually going to change `_populateResult` back to `_populateScore`. It's populating the score dom with values

my issue was that if someone was going into the codebase to see where audits' titles and descriptions are set, they probably wouldn't think it would be in `_populateScore`. If the issue is the template's id, that can change too :P

This is all going to change a lot in the near future, so if you feel strongly we can defer

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -212,19 +292,25 @@ window.ReportRenderer = class ReportRenderer {
+
+    // Append audit details to to header section so the entire audit is within one <details>.
+    const header = element.querySelector('.lighthouse-score__header');
+    if (header) {
```
it was meant to protect against `null` querySelector results like methods. But removed since we don't care about that anymore.

ebidel commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -212,19 +292,25 @@ window.ReportRenderer = class ReportRenderer {
+
+// Exports
+window.ReportRenderer = ReportRenderer;
+window.DOM = DOM;
```
Mocha needs it, but I've switched to proper module exports for Node env.

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -86,34 +85,78 @@ window.ReportRenderer = class ReportRenderer {
+  }
+
+  /**
+   * @param {!DocumentFragment|Element} element DOM node to populate with values.
```
`{!DocumentFragment|!Element}` then

ebidel replied:
done

brendankenny commented on lighthouse-core/report/v2/report-renderer.js:
```diff
@@ -213,18 +256,22 @@ window.ReportRenderer = class ReportRenderer {
+    element.appendChild(this._renderAuditScore(audit));
+
+    // Append audit details to to header section so the entire audit is within one <details>.
+    const header = element.querySelector('.lighthouse-score__header');
```
since this depends on markup introduced by `_renderAuditScore`, move header/details stuff in there?

ebidel replied:
done

brendankenny commented on lighthouse-core/report/v2/templates.html:
```diff
@@ -0,0 +1,26 @@
+      <div class="lighthouse-score__snippet">
+        <span class="lighthouse-score__title"><!-- fill me --></span>
+      </div>
+      <div class="lighthouse-score__description"><!-- fill me --></div>
```
worth putting in a comment(s) to show where audits (and details for the below template) will go?

ebidel replied:
I like the idea, but right now it's not worth it. Things are going to move around as we iterate. Don't want comments to get out of whack.

brendankenny replied:
> I like the idea, but right now it's not worth it. Things are going to move around as we iterate. Don't want comments to get out of whack.

you say that now, but we'll have to see if anyone remembers later on :)

brendankenny commented on lighthouse-core/test/report/v2/report-generator-test.js:
```diff
@@ -143,6 +146,20 @@ describe('ReportGeneratorV2', () => {
+      const page = jsdom.jsdom(new ReportGeneratorV2().generateReportHtml({}));
+      const templates = jsdom.jsdom(TEMPLATES_FILE);
+
+      assert.deepEqual(page.querySelectorAll('template[id^="tmpl-"]'),
```
`deepStrictEqual`? Deep equality checking also gets weird with e.g. empty objects and properties on prototypes ([it doesn't check them](https://nodejs.org/api/assert.html#assert_assert_deepequal_actual_expected_message)), so it may be worth doing an explicit comparison with the expected elements

ebidel replied:
done

ebidel commented:
PTAL

brendankenny commented on lighthouse-core/test/report/v2/report-generator-test.js:
```diff
@@ -143,6 +146,20 @@ describe('ReportGeneratorV2', () => {
+      const page = jsdom.jsdom(new ReportGeneratorV2().generateReportHtml({}));
+      const templates = jsdom.jsdom(TEMPLATES_FILE);
+
+      assert.deepStrictEqual(page.querySelectorAll('template[id^="tmpl-"]'),
```
just checked, `assert.deepStrictEqual` also doesn't work here (besides checking that there's the same number of templates) because `Object.keys()` on the two `HTMLTemplateElement`s returns `[]`, so any two `HTMLTemplateElement`s are considered deep equal. You'll probably need to manually check the innerHTMLs are equal or something

brendankenny replied:
could do something like
```js
const page = jsdom.jsdom(new ReportGeneratorV2().generateReportHtml({}));
const templates = jsdom.jsdom(TEMPLATES_FILE);

const pageTemplates = page.querySelectorAll('template[id^="tmpl-"]');
const rawTemplates = templates.querySelectorAll('template[id^="tmpl-"]');
rawTemplates.forEach((expectedTemplate, index) => {
  assert.strictEqual(pageTemplates[index].innerHTML, expectedTemplate.innerHTML);
});
```

ebidel replied:
switched to using lengths.

brendankenny replied:
> switched to using lengths

lulz, well the `assert.deepEqual` was checking that, but I guess at least now it's explicit
